core/metacling/src/TCling.cxx

```diff
@@ -1303,8 +1305,16 @@ TCling::TCling(const char *name, const char *title, const char* const argv[])
    // Add the overlay file. Note that we cannot factor it out for both root
    // and rootcling because rootcling activates modules only if -cxxmodule
    // flag is passed.
-   if (fCxxModulesEnabled && !fromRootCling)
+   if (fCxxModulesEnabled && !fromRootCling) {
       clingArgsStorage.push_back("-includedir_loc=" + std::string(TROOT::GetIncludeDir().Data()));
+      clingArgsStorage.push_back(("-fmodule-map-file=" +
+                                  TROOT::GetIncludeDir() + "/module.modulemap").Data());
```

Comment:
That seems quite a bit of work. I first want to fix the `/usr/local/root` issue. Fixing ROOT-9403 seems quite a bit of work. I will put it on the list but it is not yet modules critical.

Reply:
It is critical from the point of view of user. If ROOT installs `/usr/include/module.modulemap`, no distro will enable modules in their packaging (e.g. conda), and it may also interfere with other packages that need a module map if ROOT's own module map is found first. At the very least the module map should be renamed to a less generic thing, like `root.modulemap`.

Reply:
We must not install the ROOT headers in /usr/include. This is pretty terrible and it should be resolved first. I can rename the module.modulemap to root.modulemap however that would break `-Dcxxmodules=On`. We can probably install it as `root.modulemap` but that that would likely be a workaround.

Reply:
> We must not install the ROOT headers in /usr/include

IIUC "we" don't, the default gnuinstall dir is `/usr/include/root` according to [this comment](https://sft.its.cern.ch/jira/browse/ROOT-9403?focusedCommentId=106073&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-106073). So this is not as much of an issue as I initially understood it to be, and could be addresses later if actually needed. OTOH Guilherme says:

> When installing into /usr, ROOT will create /usr/include/module.modulemap

which seems to contradict the Jira comment. What am I misunderstanding?

Reply:
> We must not install the ROOT headers in /usr/include. This is pretty terrible and it should be resolved first.

That's [ROOT-9403](https://sft.its.cern.ch/jira/browse/ROOT-9403). It would indeed be nice if that got fixed.

>  I can rename the module.modulemap to root.modulemap however that would break -Dcxxmodules=On.

Why?

Reply:
> IIUC "we" don't, the default gnuinstall dir is /usr/include/root according to this comment. 
> What am I misunderstanding?

The `gnuinstall` option is disabled by default, which makes ROOT install everything into `/usr/include/` by default.

Reply:
> > We must not install the ROOT headers in /usr/include. This is pretty terrible and it should be resolved first.
> 
> That's [ROOT-9403](https://sft.its.cern.ch/jira/browse/ROOT-9403). It would indeed be nice if that got fixed.
> 
> > I can rename the module.modulemap to root.modulemap however that would break -Dcxxmodules=On.
> 
> Why?

Because -Dcxxmodules=On presumes implicit modules which presume the modulemap files to be called module.modulemap. So if ROOT is installed and somebody depends on ROOT and builds his software with -fmodules (the ROOT *compile-time* modules will not be automatically enabled). If you look at the osx distributions most of the libraries and frameworks come with a `module.modulemap` file. They do not have special names -- it is a responsibility of the installation to find a proper (isolated) place for the header files and their modulemap to live.

Reply:
I would propose to start gradually fix issues added in Jira long time ago and raised in the context of  this PR, which had initially the different purpose. Now it is important to unblock compilation of ROOT with C++ modules in case of pre-installed modulemap, such as a next case: http://cdash.cern.ch/viewBuildError.php?buildid=744995 and I am going to check if it will fix Jakob's use case. This PR will also unblock the case when user sourced dev3/dev4 builds and trying to build his custom ROOT locally.